### ISSUE TITLE
improper "Method declaration is obscured" error with singledispatchmethod

### ISSUE THREAD

**Issue by bavalpey** [microsoft/pyright#2223]
**Describe the bug**
Using singledispatchmethod, as described in [pep 443](https://www.python.org/dev/peps/pep-0443/), incorrectly reports the `Method declaration "_" is obscured by a declaration of the same name` parse error.
Honestly, I don't know if this is fixable with a static type checker.  In any case, the problem isn't trivial, as the large number of messages / errors from this clutter up other error messages.  Turning off that specific warning isn't ideal, either, since I would lose the existing functionality in all places.

**To Reproduce**
Use python's `@singledispatchmethod` decorator to register  different singledispatch methods with the same name.

**Expected behavior**
There should be no parse errors.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
try:
    from functools import singledispatchmethod
except ImportError:
    from singledispatchmethod import singledispatchmethod


@singledispatchmethod
def foo(a):
    print(f"foo does not dispatch on argument of {type(a).__name__}")

@foo.register(int)
def _(a: int): # pyright reports: Method declaration "_" is obscured by a declaration of the same name
    print(f"foo dispatches on type int!, value is: {a}")

@foo.register(float)
def _(a: float):
    print(f"foo dispatches on type float! Value is: {a}")
```

**VS Code extension or command-line**
Command line tool 1.1.162

**Comment by erictraut** [comment]
I presume that the dispatch registry mechanism doesn't care about the names of each registered function. Would it be feasible to give each registered function a unique name, like `foo_int` and `foo_float`? That would also help with debugging, since stack crawls would have unique names.

Is it considered standard practice to use `_` as the name of the single-dispatch callbacks? PEP 443 uses that convention in its examples, but it doesn't make any explicit statements about the convention. If this is pretty standard, we could simply suppress the "Function declaration is obscured" check when the name of the function is `_`.

Thoughts?

**Comment by erictraut** [comment]
Incidentally, mypy also emits the same error:
```
error: Name "_" already defined on line 10
```

**Comment by bavalpey** [comment]
> Is it considered standard practice to use `_` as the name of the single-dispatch callbacks? PEP 443 uses that convention in its examples, but it doesn't make any explicit statements about the convention. If this is pretty standard, we could simply suppress the "Function declaration is obscured" check when the name of the function is `_`.

From the examples that I've seen and how I've been using it, yes, that seems to be standard practice, as long as the registered function for the singledispatch is only ever used in that manner (e.g., the function being registered is only ever called via the dispatch).
I think suppressing the "function declaration is obscured" check for _ function names is fine, sort of like how the unused variable warning is ignored for variables named _

**Comment by erictraut** [comment]
OK, I've modified the logic to suppress this diagnostic check with the symbol name is `_`. This change will be in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.164, which I just published. It will also be in the next release of Pylance.
